parseAPI/src/IA_x86.C

```diff
@@ -91,68 +94,77 @@ bool IA_x86::isFrameSetupInsn(Instruction i) const
     return false;
 }
 
-class nopVisitor : public InstructionAPI::Visitor
-{
-    public:
-        nopVisitor() : foundReg(false), foundImm(false), foundBin(false), isNop(true) {}
-        virtual ~nopVisitor() {}
 
-        bool foundReg;
-        bool foundImm;
-        bool foundBin;
-        bool isNop;
+class leaSimplifyVisitor : public InstructionAPI::Visitor {
+        using ExprNodeValues = boost::variant2::variant<Expression*, RegisterAST*, uint64_t>;
+        using ExprStack = std::stack<ExprNodeValues>;
+        ExprStack exprStack;
 
-        virtual void visit(BinaryFunction*)
-        {
-            if (foundBin) isNop = false;
-            if (!foundImm) isNop = false;
-            if (!foundReg) isNop = false;
-            foundBin = true;
+    public:
+        void visit(BinaryFunction *bf) override {
+            auto op1 = exprStack.top();
+            exprStack.pop();
+            auto op2 = exprStack.top();
+            exprStack.pop();
+
+            auto value = boost::variant2::get_if<uint64_t>(&op1);
+            if (!value)  {
+                std::swap(op1, op2);  // op1 not an immediate, swap and try what was in op2
+                value = boost::variant2::get_if<uint64_t>(&op1);
+            }
+            if (value && ((*value == 0 && bf->isAdd()) || (*value == 1 && bf->isMultiply())))  {
+                exprStack.push(op2);  // additive or multiplicative identity, simplify to op2
+            }  else  {
+                exprStack.push(bf);   // no simplification, use BinaryFunction
+            }
         }
-        virtual void visit(Immediate *imm)
-        {
-            if (imm != 0) isNop = false;
-            foundImm = true;
+        void visit(Immediate *imm) override {
+        auto const value = imm->eval().convert<uint64_t>();
+            exprStack.push(value);
         }
-        virtual void visit(RegisterAST *)
-        {
-            foundReg = true;
+        void visit(RegisterAST *reg) override {
+            exprStack.push(reg);
         }
-        virtual void visit(Dereference *)
-        {
-            isNop = false;
+        void visit(Dereference *deref) override {
+            exprStack.push(deref);
```

Comment:
Don't we need exprStack.pop() here? It is not going to affect LEA instruction, but if someone tries to use the same visitor for other ASTs it might.

Reply:
I added a reset() method that resets the visitor to the initial state. It is unused as the iterator is never resused, but could be if there is a need to reuse the iterator.  The only other change is whitespace and a comment.

Reply:
Sorry, I misunderstood your comment. You are correct that there should be a pop. I have pushed a new version.